### PR TITLE
Fixed #33279 -- Fixed handling time zones with "-" sign in names.

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket [#33279](https://code.djangoproject.com/ticket/33279)

### CODE REVIEW

File: django/utils/dateparse.py [hunk 1]
```diff
@@ -156,3 +156,16 @@ def parse_duration(value):
...
+
+def parse_timezone_delta(tzname):
+    """
+    Parse delta time in timezone name
+
+    Return sign index and delta value.
+    """
+    for sign in ("+", "-"):
+        if sign in tzname and parse_time(tzname.split(sign)[-1]):
+            return tzname.find(sign), tzname[tzname.find(sign):]
```
Comment (felixxm) [inline django/utils/dateparse.py:169]: We could return `tzname` and `offset` instead. This should simplify `_sqlite_datetime_parse()` and `_prepare_tzname_delta()`, e.g.
```python
def split_timezone_delta(tzname):
    for sign in ('+', '-'):
        if sign in tzname:
            name, offset = tzname.rsplit(sign, 1)
            if offset and parse_time(offset):
                return name, f'{sign}{offset}'
    return tzname, None
```

**Review by felixxm — COMMENTED** [review]
@cansarigol Thanks :+1: 

Can we add a test as in the ticket description? I think it's valuable.

**Comment by cansarigol** [comment]
@felixxm I also added sign index into the response for postgresql backend. Could you have a look at it again?

File: django/utils/dateparse.py [hunk 2]
```diff
@@ -156,3 +156,17 @@ def parse_duration(value):
...
+    Parse delta time in timezone name
+
+    Return name, delta value and sign index.
+    """
+    for sign in ('+', '-'):
+        if sign in tzname:
+            name, offset = tzname.rsplit(sign, 1)
+            if offset and parse_time(offset):
+                return name, f'{sign}{offset}', tzname.find(sign)
+    return tzname, None, -1
```
Comment (felixxm) [inline django/utils/dateparse.py:172]: It's probably worth returning `sign` separately, e.g.:
```python
    for sign in ('+', '-'):
        if sign in tzname:
            name, offset = tzname.rsplit(sign, 1)
            if offset and parse_time(offset):
                return name, sign, offset
    return tzname, None, None
```
then we will have a simpler implementation on SQLite:
```python
    if tzname is not None and tzname != conn_tzname:
        tzname, sign, offset = split_timezone_delta(tzname)
        if offset:
            hours, minutes = offset.split(':')
            offset_delta = datetime.timedelta(hours=int(hours), minutes=int(minutes))
            dt += offset_delta if sign == '+' else -offset_delta
```
and PostgreSQL:
```python
        tzname, sign, offset = split_timezone_delta(tzname)
        if offset:
            sign = '-' if sign == '+' else '+'
            return f'{tzname}{sign}{offset}'
        return tzname
```
and maybe a bit more readable on MySQL and Oracle:
```python
tzname, sign, offset = split_timezone_delta(tzname)    
return f'{sign}{offset}' if offset else tzname
```
What do you think?

Comment (cansarigol) [reply django/utils/dateparse.py]: this is definitely better solution. appied it. Thanks. @felixxm

**Comment by claudep** [comment]
I guess you noticed the test failures, right :-)

**Comment by cansarigol** [comment]
> I guess you noticed the test failures, right :-)

thanks i didn't find an opportunity to fix it.

File: django/utils/dateparse.py [hunk 3]
```diff
@@ -156,3 +156,17 @@ def parse_duration(value):
         if match.re == iso8601_duration_re:
             days *= sign
         return days + sign * datetime.timedelta(**kw)
+
+
+def split_timezone_delta(tzname):
```
Comment (felixxm) [inline django/utils/dateparse.py:161]: I moved this helper to the `django.db.backends.utils`.

**Review by felixxm — COMMENTED** [review]
@cansarigol Thanks :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by claudep** [comment]
For me, this is a serious enough issue to warrant inclusion in the 4.0.x branch.

**Comment by felixxm** [comment]
> For me, this is a serious enough issue to warrant inclusion in the 4.0.x branch.

There are not many time zones that are affected but I don't see an easy workaround, so agreed.

**Pull request merged**